### PR TITLE
Re-organize shared src packages so we can do true IVT between src ass…

### PR DESCRIPTION
**Pull request by pranavkm**
…emblies

### CODE REVIEW

File: src/Microsoft.AspNetCore.Mvc.Abstractions/Properties/AssemblyInfo.cs [hunk 1]
```diff
@@ -3,5 +3,36 @@
 
 using System.Runtime.CompilerServices;
 
+[assembly: InternalsVisibleTo("Microsoft.AspNetCore.Mvc, PublicKey=0024000004800000940000000602000000240000525341310004000001000100f33a29044fa9d740c9b3213a93e57c84b472c84e0b8a0e1ae48e67a9f8f6de9d5f7f3d52ac23e48ac51801f1dc950abe901da34d2a9e3baadb141a17c77ef3c565dd5ee5054b91cf63bb3c6ab83f72ab3aafe93d0fc3c2348b764fafb0b1c0733de51459aeab46580384bf9d74c4e28164b7cde247f891ba07891c9d872ad2bb")]
+[assembly: InternalsVisibleTo("Microsoft.AspNetCore.Mvc.Abstractions, PublicKey=0024000004800000940000000602000000240000525341310004000001000100f33a29044fa9d740c9b3213a93e57c84b472c84e0b8a0e1ae48e67a9f8f6de9d5f7f3d52ac23e48ac51801f1dc950abe901da34d2a9e3baadb141a17c77ef3c565dd5ee5054b91cf63bb3c6ab83f72ab3aafe93d0fc3c2348b764fafb0b1c0733de51459aeab46580384bf9d74c4e28164b7cde247f891ba07891c9d872ad2bb")]
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.Abstractions/Properties/AssemblyInfo.cs]: Why would `Microsoft.AspNetCore.Mvc.Abstractions` share `internal`s with itself?

File: src/Microsoft.AspNetCore.Mvc.Core/Properties/AssemblyInfo.cs [hunk 2]
```diff
@@ -4,8 +4,38 @@
...
+[assembly: InternalsVisibleTo("Microsoft.AspNetCore.Mvc.ApiExplorer, PublicKey=0024000004800000940000000602000000240000525341310004000001000100f33a29044fa9d740c9b3213a93e57c84b472c84e0b8a0e1ae48e67a9f8f6de9d5f7f3d52ac23e48ac51801f1dc950abe901da34d2a9e3baadb141a17c77ef3c565dd5ee5054b91cf63bb3c6ab83f72ab3aafe93d0fc3c2348b764fafb0b1c0733de51459aeab46580384bf9d74c4e28164b7cde247f891ba07891c9d872ad2bb")]
+[assembly: InternalsVisibleTo("Microsoft.AspNetCore.Mvc.Core, PublicKey=0024000004800000940000000602000000240000525341310004000001000100f33a29044fa9d740c9b3213a93e57c84b472c84e0b8a0e1ae48e67a9f8f6de9d5f7f3d52ac23e48ac51801f1dc950abe901da34d2a9e3baadb141a17c77ef3c565dd5ee5054b91cf63bb3c6ab83f72ab3aafe93d0fc3c2348b764fafb0b1c0733de51459aeab46580384bf9d74c4e28164b7cde247f891ba07891c9d872ad2bb")]
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.Core/Properties/AssemblyInfo.cs]: Why would `Microsoft.AspNetCore.Mvc.Core` share `internal`s with itself or `Microsoft.AspNetCore.Mvc.Abstractions`? Same for all later .csproj files.

File: src/Microsoft.AspNetCore.Mvc.Core/Microsoft.AspNetCore.Mvc.Core.csproj [hunk 3]
```diff
@@ -11,6 +11,7 @@ Microsoft.AspNetCore.Mvc.FromFormAttribute
 Microsoft.AspNetCore.Mvc.RequireHttpsAttribute
 Microsoft.AspNetCore.Mvc.RouteAttribute</Description>
     <TargetFramework>netstandard2.0</TargetFramework>
+    <RootNamespace>Microsoft.AspNetCore.Mvc</RootNamespace>
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.Core/Microsoft.AspNetCore.Mvc.Core.csproj]: How is this change (which brings in a lot of additional files) related to the IVT one? In addition, the ambiguous `Resources` references you had to clean up get quite hideous e.g. `AspNetCore.Mvc.RazorPages.Resources.ArgumentCannotBeNullOrEmpty`.

**Comment by pranavkm** [comment]
🆙 📅

**Comment by pranavkm** [comment]
Merging this in later today if there aren't any additional concerns.

**Comment by davidfowl** [comment]
I hate it, but carry on

**Comment by rynowak** [comment]
My concern is still that I'm not sure what the next steps are. I think this is premature without understanding what we will **do** with it.

**Comment by pranavkm** [comment]
I based my changes for ViewContext.ViewData \ PageContext.ViewData to use `internal` interfaces - https://github.com/aspnet/Mvc/pull/8071/files#diff-a76799dccb08c23740d2eead57c49327R5 and internal properties - https://github.com/aspnet/Mvc/pull/8071/files#diff-2cd993346d0e7f84012a35068ad131c0R99. Ideally we just have a better design for either of those, but doing this PR gets us out of the dilemma of creating throw away types until we have a better solution.

**Comment by rynowak** [comment]
Right, so what will **do** next? Is there a follow up action? Are there planned areas to refactor that have been understood and prioritized?

**Comment by pranavkm** [comment]
> Are there planned areas to refactor that have been understood and prioritized

I wanted to figure out if we want to actively go and update existing `.Internal` types -> `internal` or do them passively as and when we touch them. We've taken the latter approach to make current `.Internal` types -> `internal`, but there's no reason we couldn't spend a bit of time and clean up the repo. We could have a chat about it next week?

File: src/Microsoft.AspNetCore.Mvc.Core/Properties/AssemblyInfo.cs [hunk 4]
```diff
@@ -4,8 +4,36 @@
 using System.Runtime.CompilerServices;
 using Microsoft.AspNetCore.Mvc.Formatters;
 
-[assembly: InternalsVisibleTo("Microsoft.AspNetCore.Mvc.Test, PublicKey=0024000004800000940000000602000000240000525341310004000001000100f33a29044fa9d740c9b3213a93e57c84b472c84e0b8a0e1ae48e67a9f8f6de9d5f7f3d52ac23e48ac51801f1dc950abe901da34d2a9e3baadb141a17c77ef3c565dd5ee5054b91cf63bb3c6ab83f72ab3aafe93d0fc3c2348b764fafb0b1c0733de51459aeab46580384bf9d74c4e28164b7cde247f891ba07891c9d872ad2bb")]
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.Core/Properties/AssemblyInfo.cs]: Should leave out these additional IVT changes until and unless we change pubternal classes in Core.

File: src/Microsoft.AspNetCore.Mvc.Razor/Properties/AssemblyInfo.cs [hunk 5]
```diff
@@ -3,5 +3,7 @@
 
 using System.Runtime.CompilerServices;
 
+[assembly: InternalsVisibleTo("Microsoft.AspNetCore.Mvc.RazorPages, PublicKey=0024000004800000940000000602000000240000525341310004000001000100f33a29044fa9d740c9b3213a93e57c84b472c84e0b8a0e1ae48e67a9f8f6de9d5f7f3d52ac23e48ac51801f1dc950abe901da34d2a9e3baadb141a17c77ef3c565dd5ee5054b91cf63bb3c6ab83f72ab3aafe93d0fc3c2348b764fafb0b1c0733de51459aeab46580384bf9d74c4e28164b7cde247f891ba07891c9d872ad2bb")]
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.Razor/Properties/AssemblyInfo.cs]: Should leave out these additional IVT changes until and unless we change pubternal classes in Mvc.Razor.

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/AttributeRoute.cs [hunk 6]
```diff
@@ -178,12 +178,12 @@ private static List<RouteInfo> GetRouteInfos(IReadOnlyList<ActionDescriptor> act
                 var allErrors = string.Join(
                     Environment.NewLine + Environment.NewLine,
                     errors.Select(
-                        e => Resources.FormatAttributeRoute_IndividualErrorMessage(
+                        e => Core.Resources.FormatAttributeRoute_IndividualErrorMessage(
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/AttributeRoute.cs]: Recommend `using Resources = Core.Resources;` (or similar) wherever these namespace qualifications are needed more than once per class.

File: src/Microsoft.AspNetCore.Mvc.Razor/Properties/AssemblyInfo.cs (hunk 5 above)
Comment (pranavkm) [reply src/Microsoft.AspNetCore.Mvc.Razor/Properties/AssemblyInfo.cs]: Need this in a follow up PR - https://github.com/aspnet/Mvc/pull/8071

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/AttributeRoute.cs (hunk 6 above)
Comment (pranavkm) [reply src/Microsoft.AspNetCore.Mvc.Core/Internal/AttributeRoute.cs]: Sure

File: src/Microsoft.AspNetCore.Mvc.Core/Properties/AssemblyInfo.cs (hunk 4 above)
Comment (pranavkm) [reply src/Microsoft.AspNetCore.Mvc.Core/Properties/AssemblyInfo.cs]: Speaking to @rynowak offline, we landed on all of the shared src packages to Core from Abstractions (the ones that Abstractions doesn't use). So we kinda need these now.

File: src/Microsoft.AspNetCore.Mvc.Razor/Properties/AssemblyInfo.cs (hunk 5 above)
Comment (dougbu) [reply src/Microsoft.AspNetCore.Mvc.Razor/Properties/AssemblyInfo.cs]: Then do this in #8071

File: src/Microsoft.AspNetCore.Mvc.Core/Properties/AssemblyInfo.cs (hunk 4 above)
Comment (dougbu) [reply src/Microsoft.AspNetCore.Mvc.Core/Properties/AssemblyInfo.cs]: If Abstractions shares everything with everyone, why do we need additional IVTs in core?

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionInvoker.cs [hunk 7]
```diff
@@ -312,7 +312,7 @@ private Task Next(ref State next, ref Scope scope, ref object state, ref bool is
             if (_actionExecutingContext.Result != null)
             {
                 // If we get here, it means that an async filter set a result AND called next(). This is forbidden.
-                var message = Resources.FormatAsyncActionFilter_InvalidShortCircuit(
+                var message = Core.Resources.FormatAsyncActionFilter_InvalidShortCircuit(
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionInvoker.cs]: Redundant namespace qualification.

File: src/Microsoft.AspNetCore.Mvc.Razor/Properties/AssemblyInfo.cs (hunk 5 above)
Comment (pranavkm) [reply src/Microsoft.AspNetCore.Mvc.Razor/Properties/AssemblyInfo.cs]: Leaving as is.

File: src/Microsoft.AspNetCore.Mvc.Core/Properties/AssemblyInfo.cs (hunk 4 above)
Comment (dougbu) [reply src/Microsoft.AspNetCore.Mvc.Core/Properties/AssemblyInfo.cs]: Oh, I get it now.

File: src/Microsoft.AspNetCore.Mvc.RazorPages/DependencyInjection/MvcRazorPagesMvcBuilderExtensions.cs [hunk 8]
```diff
@@ -52,12 +52,12 @@ public static IMvcBuilder WithRazorPagesRoot(this IMvcBuilder builder, string ro
 
             if (string.IsNullOrEmpty(rootDirectory))
             {
-                throw new ArgumentException(Resources.ArgumentCannotBeNullOrEmpty, nameof(rootDirectory));
+                throw new ArgumentException(AspNetCore.Mvc.RazorPages.Resources.ArgumentCannotBeNullOrEmpty, nameof(rootDirectory));
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/DependencyInjection/MvcRazorPagesMvcBuilderExtensions.cs]: Use `using` aliases from here to end of PR

**Pull request closed without merging**